app/models/agents/shell_command_agent.rb

```diff
@@ -87,24 +98,33 @@ def handle(opts, event = nil)
       end
     end
 
-    def run_command(path, command)
-      result = nil
-      errors = nil
-      exit_status = nil
-
-      Dir.chdir(path){
-        begin
-          stdin, stdout, stderr, wait_thr = Open3.popen3(command)
-          exit_status = wait_thr.value.to_i
-          result = stdout.gets(nil)
-          errors = stderr.gets(nil)
-        rescue Exception => e
-          errors = e.to_s
+    def run_command(path, command, stdin)
+      begin
+        rout, wout = IO.pipe
+        rerr, werr = IO.pipe
+        rin,  win = IO.pipe
+
+        pid = spawn(*command, chdir: path, out: wout, err: werr, in: rin)
+
+        wout.close
+        werr.close
+        rin.close
+
+        if stdin
+          win.write stdin
+          win.close
         end
-      }
 
-      result = result.to_s.strip
-      errors = errors.to_s.strip
+        (result = rout.read).strip!
+        (errors = rerr.read).strip!
+
+        _, status = Process.wait2(pid)
+        exit_status = status.exitstatus
+      rescue Exception => e
```

Comment:
SystemExit and Interrupt are not likely worth rescuing, so I've changed it as you suggested.